tests/std/tests/P2231R1_complete_constexpr_optional_variant/test.cpp

```diff
@@ -0,0 +1,334 @@
+    constexpr With_nontrivial_destructor(const int val) noexcept : _val(val) {}
+    constexpr With_nontrivial_destructor(initializer_list<int> vals) noexcept : _val(*vals.begin()) {}
+    constexpr ~With_nontrivial_destructor() {}
+
+    constexpr bool operator==(const int right) const noexcept {
+        return _val == right;
+    }
+};
+
+struct Dummy {
+    int _val = 0;
+};
+
+template <class T>
+constexpr bool test_optional() {
+    { // empty construction
+        optional<T> default_constructed;
+        assert(!default_constructed.has_value());
+
+        optional<T> nullopt_constructed{nullopt};
+        assert(!nullopt_constructed.has_value());
+    }
+
+    { // construction from underlying type
+        const T input{42};
+        optional<T> copy_constructed{input};
+        assert(copy_constructed.has_value());
+        assert(*copy_constructed == 42);
+
+        optional<T> move_constructed{T{42}};
+        assert(move_constructed.has_value());
+        assert(*move_constructed == 42);
+
+        optional<T> copy_assigned;
+        assert(!copy_assigned.has_value());
+        copy_assigned = input;
+        assert(copy_assigned.has_value());
+        assert(*copy_assigned == 42);
+
+        optional<T> move_assigned;
+        assert(!move_assigned.has_value());
+        move_assigned = T{42};
+        assert(move_assigned.has_value());
+        assert(*move_assigned == 42);
+    }
+
+    { // construction from convertible type
+        const int input{42};
+        optional<T> copy_constructed{input};
+        assert(copy_constructed.has_value());
+        assert(*copy_constructed == 42);
+
+        optional<T> move_constructed{42};
+        assert(move_constructed.has_value());
+        assert(*move_constructed == 42);
+
+        optional<T> copy_assigned;
+        assert(!copy_assigned.has_value());
+        copy_assigned = input;
+        assert(copy_assigned.has_value());
+        assert(*copy_assigned == 42);
+
+        optional<T> move_assigned;
+        assert(!move_assigned.has_value());
+        move_assigned = 42;
+        assert(move_assigned.has_value());
+        assert(*move_assigned == 42);
+    }
+
+    { // construction from optional with same type
+        optional<T> constructed{42};
+        assert(constructed.has_value());
+        assert(*constructed == 42);
+
+        optional<T> copy_constructed{constructed};
+        assert(copy_constructed.has_value());
+        assert(*copy_constructed == 42);
+
+        optional<T> move_constructed{move(constructed)};
+        assert(move_constructed.has_value());
+        assert(*move_constructed == 42);
+
+#if !(defined(__clang__) || defined(__EDG__)) // TRANSITION, DevCom-1331017
+        if constexpr (!is_same_v<T, With_nontrivial_destructor>) {
+            if (!is_constant_evaluated()) {
+#endif // TRANSITION
+                optional<T> copy_assigned;
+                assert(!copy_assigned.has_value());
+                copy_assigned = constructed;

```

Comment:
I observe that `constructed` has been moved-from above, but the Standard guarantees that it still `has_value()`, and the types being tested are unchanged when they're moved-from, so this is well-defined. (Ditto for the `variant` tests below.) No change requested.

Reply:
Changed so that we properly use first `move_constructed` and then `copy_assigned`